### PR TITLE
Handle ICProxy addresses during gpexpand

### PR DESCRIPTION
**Pull request by interma**
Current gpexpand doesn't handle the ICProxy addresses.
So, if we are in icproxy mode: the proxy addresses of newly added segments are still empty after expansion. This causes queries to hang forever.
        
We have to check if the user have already set `gp_interconnect_proxy_addresses` before and prompt them to set a proper value via `gpconfig` before expansion.

The new logic in my PR:
* Add a new step (`gpexpand.validate_icproxy_addr()`) to check the icproxy address in `gpexpand -i {input_file}` mode
    * If `gp_interconnect_proxy_addresses` is not set, nothing needs to do and just skip the following operations
    * If set, loop the new seg info in the `{input_file}`, to check if its dbid exists in `gp_interconnect_proxy_addresses`
        * If all are in it, nothing needs to do
        * If any dbid doesn't exist, raise an ERROR: prompt a gpconfig command template to let user set the missing icproxy addresses and rerun gpexpand again (if `gp_interconnect_type != proxy`, just raise a WARN with the same message)
* (other same steps as before, don't change them)

Example outputs:
```
[INFO]:-ICProxy addresses has been set to: 1:-1:gpdb:1432,2:0:gpdb:2000,3:1:gpdb:2001, start to checking if it contains newly added segs
[WARNING]:- Missing the address of dbid 4 in gp_interconnect_proxy_addresses
[WARNING]:- Missing the address of dbid 5 in gp_interconnect_proxy_addresses
[ERROR]:-Please run gpconfig to set gp_interconnect_proxy_addresses (replace the {port} with a proper value):
[ERROR]:- gpconfig -c gp_interconnect_proxy_addresses -v "'1:-1:gpdb:1432,2:0:gpdb:2000,3:1:gpdb:2001,4:2:gpdb:{port},5:3:gpdb:{port}'" --skipvalidation
[ERROR]:-then rerun "gpexpand -i {input_file}" .
```

I think the above steps are good enough for gpexpand: simple and serves as a reminder to user (when they don't notice cluster is in icproxy mode).

## Here are some reminders before you submit the pull request
- [x] Add tests for the change
- [ ] Document changes
- [x] Communicate in the mailing list if needed
- [x] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: gpMgmt/bin/gpexpand [hunk 1]
```diff
@@ -1999,6 +1999,53 @@ class gpexpand:
...
+                    self.logger.error("Please run gpconfig to set gp_interconnect_proxy_addresses (replace the {port} with a proper value):")
+                    self.logger.error(" gpconfig -c gp_interconnect_proxy_addresses -v \"'%s'\" --skipvalidation" % (new_addresses_string))
+                    self.logger.error('then rerun "gpexpand -i {input_file}" .')
+                else:
+                    self.logger.warning("Recommended that run gpconfig to set gp_interconnect_proxy_addresses (replace the {port} with a proper value):")
+                    self.logger.warning(" gpconfig -c gp_interconnect_proxy_addresses -v \"'%s'\" --skipvalidation" % (new_addresses_string))
+            else:
+                self.logger.info("Checking ICProxy addresses done: no need to change")
+
+        raise Exception("@interma let's stop here")
```
Comment (interma) [inline gpMgmt/bin/gpexpand:2048]: Just for test, I will remove it soon.

**Comment by interma** [comment]
Some jobs of dev pipeline failed, trying to fix them.

**Review by adam8157 — COMMENTED** [review]
Is it possible to have a test in `gpMgmt/test/behave/mgmt_utils/gpexpand.feature`?

File: gpMgmt/bin/gpexpand [hunk 2]
```diff
@@ -1999,6 +1999,54 @@ class gpexpand:
...
+                if seg.dbid not in dbid_set:
+                    need_prompt = True
+                    new_addresses_string += ",%s:%s:%s:{port}" % (seg.dbid, seg.contentId, seg.hostname.strip())
+                    self.logger.warning(" Missing the address of dbid %s in gp_interconnect_proxy_addresses" % (seg.dbid))
+
+            # prompt user to use gpconfig to set the proper value of gp_interconnect_proxy_addresses
+            if need_prompt:
+                if ic_type == "proxy":
+                    self.logger.error("Please run gpconfig to set gp_interconnect_proxy_addresses (replace the {port} with a proper value):")
+                    self.logger.error(" gpconfig -c gp_interconnect_proxy_addresses -v \"'%s'\" --skipvalidation" % (new_addresses_string))
```
Comment (adam8157) [inline gpMgmt/bin/gpexpand:2040]: This is supposed to be done before rerunning gpexpand, so it's fine to have a value with more nodes than the actual, is it? For example, a 3-node cluster having a 5-nodes GUC value.

**Review by adam8157 — COMMENTED** [review]
Does gpexpand work fine with the IC proxy setup? Especially while the new segments initiating.

File: gpMgmt/bin/gpexpand (hunk 2 above)
Comment (interma) [reply gpMgmt/bin/gpexpand:2040]: Yes, I have tested it before. It's no problem.
And verified in related code:
* it only loops all addresses and set itself by dbid
https://github.com/greenplum-db/gpdb/blob/8b8d2a4630e0010215cfafc028b79caec617fb16/src/backend/cdb/motion/ic_proxy_addr.c#L313-L314
* connect to other (non-exist) peers: just write a log (only in VERBOSE mode) when failed, also no side effects.

**Comment by interma** [comment]
> Is it possible to have a test in gpMgmt/test/behave/mgmt_utils/gpexpand.feature?

Thank you, I will take a look at these test code and try to write one for my PR.

> Does gpexpand work fine with the IC proxy setup? Especially while the new segments initiating.

My test env's ICType is always set to proxy, and it's ok. I will go through the whole process of gpexpand later.

**Comment by interma** [comment]
Added behave test cases.

**Review by adam8157 — APPROVED** [review]

**Comment by interma** [comment]
Dev pipeline passed: 
https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/gpdb-dev-gpexpand_icproxy-rhel8

**Pull request merged**